pkg/ipcache/types_test.go

```diff
@@ -59,3 +66,225 @@ func Test_sortedByResourceIDsAndSource(t *testing.T) {
+
+	flagsT := ipcacheTypes.EndpointFlags{}
+	flagsT.SetSkipTunnel(true)
+
+	flagsF := ipcacheTypes.EndpointFlags{}
+	flagsF.SetSkipTunnel(false)
+
+	tests := []struct {
+		before   *resourceInfo
+		after    *resourceInfo
+		expected *resourceInfo
+	}{
+		{
+			before: &resourceInfo{
+				source: source.Local,
+				labels: labels.NewLabelsFromSortedList("source:a=b;source:k=v;"),
+			},
+			after: &resourceInfo{
+				source: source.Generated,
+				labels: labels.NewLabelsFromSortedList("source:c=d;source:k=x"),
+			},
+			expected: &resourceInfo{
+				source: source.Local,
+				labels: labels.NewLabelsFromSortedList("source:a=b;source:c=d;source:k=v"),
+			},
+		},
+		{
+			before: &resourceInfo{
+				source: source.Local,
+				labels: labels.NewLabelsFromSortedList("source:a=b;source:k=v;"),
+			},
+			after: &resourceInfo{
+				source:           source.Generated,
+				labels:           labels.NewLabelsFromSortedList("source:c=d;source:k=x"),
+				identityOverride: true,
+			},
+			expected: &resourceInfo{
+				source:           source.Local,
+				labels:           labels.NewLabelsFromSortedList("source:c=d;source:k=x"),
+				identityOverride: true,
+			},
+		},
+		// all conflicts
+		{
+			before: &resourceInfo{
+				source:            source.Local,
+				tunnelPeer:        types.TunnelPeer{Addr: ipA},
+				encryptKey:        types.EncryptKey(1),
+				requestedIdentity: types.RequestedIdentity(10),
+				endpointFlags:     flagsF,
+			},
+			after: &resourceInfo{
+				source:            source.Generated,
+				tunnelPeer:        types.TunnelPeer{Addr: ipB},
+				encryptKey:        types.EncryptKey(2),
+				requestedIdentity: types.RequestedIdentity(11),
+				endpointFlags:     flagsT,
+			},
+			expected: &resourceInfo{
+				source:            source.Local,
+				tunnelPeer:        types.TunnelPeer{Addr: ipA},
+				encryptKey:        types.EncryptKey(1),
+				requestedIdentity: types.RequestedIdentity(10),
+				endpointFlags:     flagsF,
+			},
+		},
+		// half and half, no conflicts
+		{
+			before: &resourceInfo{
+				source:     source.Local,
+				tunnelPeer: types.TunnelPeer{Addr: ipA},
+				encryptKey: types.EncryptKey(1),
+			},
+			after: &resourceInfo{
+				requestedIdentity: types.RequestedIdentity(11),
+				endpointFlags:     flagsT,
+				labels:            labels.NewLabelsFromSortedList("source:a=b;source:k=v;"),
+			},
+			expected: &resourceInfo{
+				labels:            labels.NewLabelsFromSortedList("source:a=b;source:k=v;"),
+				source:            source.Local,
+				tunnelPeer:        types.TunnelPeer{Addr: ipA},
+				encryptKey:        types.EncryptKey(1),
+				requestedIdentity: types.RequestedIdentity(11),
+				endpointFlags:     flagsT,
+			},
+		},
+	}
+
+	for i, test := range tests {
+		t.Run(strconv.Itoa(i), func(t *testing.T) {
+			logger := hivetest.Logger(t)
+			logger.Info("warnings are expected here")
+			pi := newPrefixInfo()
+
+			pi.byResource["before"] = test.before
+			pi.byResource["after"] = test.after
```

Comment:
nit: The use of "before" and "after" here may confuse future readers, since the ResourceID concept in `pkg/ipcache` is not about time or change but about multiple conflicting sources of truth being de-conflicted by the `flatten()` call. I think it would reinforce a more accurate view of the corresponding code if the test cases just used `resourceA`, `resourceB` or even more use-case oriented cases like using actual resource IDs like `cnp/cidr-policy-a` or `daemon/fqdn-name-manager/cilium.io` or such.